pandar_pointcloud/src/lib/decoder/pandar40_decoder.cpp

```diff
@@ -86,81 +94,100 @@ void Pandar40Decoder::unpack(const pandar_msgs::PandarPacket& raw_packet)
-    const auto& unit = block.units[unit_id];
-    // skip invalid points
-    if (unit.distance <= 0.1 || unit.distance > 200.0) {
-      continue;
-    }
-    double xyDistance = unit.distance * cosf(deg2rad(elev_angle_[unit_id]));
+  point.x = static_cast<float>(
+      xyDistance * sinf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
+  point.y = static_cast<float>(
+      xyDistance * cosf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
+  point.z = static_cast<float>(unit.distance * sinf(deg2rad(elev_angle_[unit_id])));
 
-    point.x = static_cast<float>(
-        xyDistance * sinf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
-    point.y = static_cast<float>(
-        xyDistance * cosf(deg2rad(azimuth_offset_[unit_id] + (static_cast<double>(block.azimuth)) / 100.0)));
-    point.z = static_cast<float>(unit.distance * sinf(deg2rad(elev_angle_[unit_id])));
+  point.intensity = unit.intensity;
+  point.distance = unit.distance;
+  point.ring = unit_id;
+  point.azimuth = block.azimuth + round(azimuth_offset_[unit_id] * 100.0f);
+  point.return_type = return_type;
+  point.time_stamp = unix_second + (static_cast<double>(packet_.usec)) / 1000000.0;
+  point.time_stamp -= (static_cast<double>(block_offset_dual_[block_id] + firing_offset_[unit_id]) / 1000000.0f);
 
-    point.intensity = unit.intensity;
-    point.distance = unit.distance;
-    point.ring = unit_id;
-    point.azimuth = block.azimuth + round(azimuth_offset_[unit_id] * 100.0f);
+  return point;
+}
 
-    point.time_stamp = unix_second + (static_cast<double>(packet_.usec)) / 1000000.0;
-    point.time_stamp -= (static_cast<double>(block_offset_single_[block_id] + firing_offset_[unit_id]) / 1000000.0f);
+PointcloudXYZIRADT Pandar40Decoder::convert(int block_id)
+{
+  PointcloudXYZIRADT block_pc(new pcl::PointCloud<PointXYZIRADT>);
 
-    block_pc->push_back(point);
+  for (auto unit_id : firing_order_) {
+    block_pc->push_back(build_point(block_id, unit_id, (packet_.return_mode == STRONGEST_RETURN) ? ReturnType::SINGLE_STRONGEST : ReturnType::SINGLE_LAST)); 
   }
   return block_pc;
 }
 
 PointcloudXYZIRADT Pandar40Decoder::convert_dual(int block_id)
 {
   //   Under the Dual Return mode, the measurements from each round of firing are stored in two adjacent blocks:
-  // · The odd number block is the last return, and the even number block is the strongest return
-  // · If the last and strongest returns coincide, the second strongest return will be placed in the even number block
+  // · The even number block is the last return, and the odd number block is the strongest return
+  // · If the last and strongest returns coincide, the second strongest return will be placed in the odd number block
   // · The Azimuth changes every two blocks
+  // · Important note: Hesai datasheet block numbering starts from 0, not 1, so odd/even are reversed here
```

Comment:
Just want to make sure if you have seen the actual points and checked these facts. (I've never checked it myself.)

Reply:
Yes, I triple checked this as it is confusing. The data reflects the above.